### PR TITLE
[Issue-#8] Sync merge

### PR DESCRIPTION
**Pull request by olehhaliak**
This PR contains some resources that could be useful for other tickets:
- exception package
- annotation package
- utility package
- minor interface updates
- jacoco config updates

Test cases will be refactored with the next PR

### CODE REVIEW

File: src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Column.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.bobocode.blyznytsia.bibernate.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
```
Comment (mvoronov) [inline src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Column.java:8]: Related to all annotations: 
 - do we need to add javadoc at this point ?
 - do we need to add `@Documented` ?

File: src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Column.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+@Target(ElementType.FIELD)
+public @interface Column {
+  String name();
```
Comment (mvoronov) [inline src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Column.java:11]: is `name` really required  or we allow to have it default value?

File: src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Table.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface Table {
+  String name();
```
Comment (mvoronov) [inline src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Table.java:11]: is `name` really required  or we allow to have it default value?

File: src/main/java/com/bobocode/blyznytsia/bibernate/entity/EntityDefinition.java
Comment (mvoronov) [inline src/main/java/com/bobocode/blyznytsia/bibernate/entity/EntityDefinition.java:1]: I am out of scope a bit, why did you delete all classes from entity package?

File: src/main/java/com/bobocode/blyznytsia/bibernate/exception/BibernateException.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.bobocode.blyznytsia.bibernate.exception;
+
+public class BibernateException extends RuntimeException {
```
Comment (mvoronov) [inline src/main/java/com/bobocode/blyznytsia/bibernate/exception/BibernateException.java:3]: I see, that javadoc is missing in all classes, do we need to add it at this point, or it will be added later ?

File: src/main/java/com/bobocode/blyznytsia/bibernate/exception/BibernateException.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  public BibernateException(String message) {
```
Comment (mvoronov) [inline src/main/java/com/bobocode/blyznytsia/bibernate/exception/BibernateException.java:8]: Please, reorder constructors, so that constructor with more params goes first (like you did in PersistenceException)

File: src/main/java/com/bobocode/blyznytsia/bibernate/persister/EntityPersister.java [hunk 6]
```diff
@@ -1,4 +1,19 @@
...
+
+  <T> Optional<T> findOneBy(Class<T> entityType, Field key, Object value);
+
+  <T> List<T> findAll(Class<T> entityType, Field key, Object value);
+
+  <T> T insert(T entity);
+
+  <T> T update(T entity);
+
+  <T> T delete(T entity);
```
Comment (mvoronov) [inline src/main/java/com/bobocode/blyznytsia/bibernate/persister/EntityPersister.java]: shouldn't it be void ?

File: src/main/java/com/bobocode/blyznytsia/bibernate/util/EntityUtil.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+public class EntityUtil {
+  public static String resolveEntityTableName(Class<?> entityType) {
+    return Optional
+        .ofNullable(entityType.getAnnotation(Table.class))
+        .map(Table::name)
+        .orElse(camelToSnakeCase(entityType.getSimpleName()));
+  }
+
+  public static Field resolveEntityIdField(Class<?> entityType) {
+    return Arrays.stream(entityType.getDeclaredFields())
```
Comment (mvoronov) [inline src/main/java/com/bobocode/blyznytsia/bibernate/util/EntityUtil.java]: Should we respect case when there are two fields annotated with `@Id` ?

File: src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/AnnotatedSampleEntity.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package com.bobocode.blyznytsia.bibernate.testdata.entity;
+
+import com.bobocode.blyznytsia.bibernate.annotation.Column;
+import com.bobocode.blyznytsia.bibernate.annotation.Id;
+import com.bobocode.blyznytsia.bibernate.annotation.Table;
+
+@Table(name = "custom_entity_table_name")
+public class AnnotatedSampleEntity {
+  @Id
+  Long id;
```
Comment (mvoronov) [inline src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/AnnotatedSampleEntity.java]: Should fields be private?

File: src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/EntityWithNoId.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package com.bobocode.blyznytsia.bibernate.testdata.entity;
+
+import com.bobocode.blyznytsia.bibernate.annotation.Entity;
+
+@Entity
+public class EntityWithNoId {
```
Comment (mvoronov) [inline src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/EntityWithNoId.java]: Would `EntityWithoutId` name be better?

File: src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/EntityWithNoId.java [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+  // @Id annotation is missing on purpose
+  Long id;
```
Comment (mvoronov) [inline src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/EntityWithNoId.java]: private ?

File: src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/SampleEntity.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Entity
+@Data
+@AllArgsConstructor
+public class SampleEntity {
+  @Id
+  Long id;
+  String someValue;
```
Comment (mvoronov) [inline src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/SampleEntity.java]: private ?

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    assertEquals("field_uppercase_abbreviation", CaseUtil.camelToSnakeCase("fieldUppercaseABBREVIATION"));
+  }
+
+  void camelToSnakeCaseConvertingClassNames(){
+    assertEquals("some_class",                   CaseUtil.camelToSnakeCase("SomeClass"));
+    assertEquals("some_class_but_longer",        CaseUtil.camelToSnakeCase("SomeClassButLonger"));
+    assertEquals("class_uppercase_abbreviation", CaseUtil.camelToSnakeCase("ClassUppercaseABBREVIATION"));
+  }
+
+}
```
Comment (mvoronov) [inline src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java]: Please, add empty line at the end of the class

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+class CaseUtilTest {
+  @Test
+  void camelToSnakeCaseConvertingFieldNames() {
+    assertEquals("some_field",                   CaseUtil.camelToSnakeCase("someField"));
+    assertEquals("some_field_but_longer",        CaseUtil.camelToSnakeCase("someFieldButLonger"));
+    assertEquals("field_uppercase_abbreviation", CaseUtil.camelToSnakeCase("fieldUppercaseABBREVIATION"));
+  }
+
+  void camelToSnakeCaseConvertingClassNames(){
```
Comment (mvoronov) [inline src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java:16]: Missing `@Test` annotation?

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
+package com.bobocode.blyznytsia.bibernate.util;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.jupiter.api.Test;
+
+class CaseUtilTest {
+  @Test
```
Comment (mvoronov) [inline src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java:8]: I would prefer to use Parametrized test

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/EntityUtilTest.java [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  @Test
+  void resolveEntityIdFieldShouldThrowMalformedEntityException() {
+    assertThrows(MalformedEntityException.class, () -> EntityUtil.resolveEntityIdField(EntityWithNoId.class));
+  }
+
+  @Test
+  void getEntityNonIdFields() {
+    var expectedField = Arrays.stream(EntityWIthMultipleNonIdFields.class.getDeclaredFields())
```
Comment (mvoronov) [inline src/test/java/com/bobocode/blyznytsia/bibernate/util/EntityUtilTest.java]: Please, rename to `expectedFields`

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/EntityUtilTest.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        .filter(field -> ! field.getName().equals("id"))
+        .toList();
+    var actualNonIdFields = EntityUtil.getEntityNonIdFields(new EntityWIthMultipleNonIdFields());
+    assertEquals(expectedField, actualNonIdFields);
+  }
+}
```
Comment (mvoronov) [inline src/test/java/com/bobocode/blyznytsia/bibernate/util/EntityUtilTest.java]: Please, add empty line at the end of the class

File: lombok.config [hunk 17]
```diff
@@ -0,0 +1,2 @@
+config.stopBubbling = true
+lombok.addLombokGeneratedAnnotation = true
```
Comment (mvoronov) [inline lombok.config]: Please, add empty line at the end of the file

**Review by mvoronov — CHANGES_REQUESTED** [review]

File: src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Column.java (hunk 1 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Column.java:8]: Javadoc will be added with the next PR, the purpose of this PR is to deblock other tickets and reduce merge conflicts

File: src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Column.java (hunk 2 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Column.java:11]: Given this is the only use-case for this annotation so far, it makes sense to require it.

File: src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Table.java (hunk 3 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Table.java:11]: Same as with Column, currently this annotation is only intended to specify a custom table name

File: src/main/java/com/bobocode/blyznytsia/bibernate/entity/EntityDefinition.java
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/entity/EntityDefinition.java:1]: EntityDefinition** classes were moved to the additional scope and will very likely be implemented differently (if they will).

File: src/main/java/com/bobocode/blyznytsia/bibernate/exception/BibernateException.java (hunk 4 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/exception/BibernateException.java:3]: Javadocs coming soon)

File: src/main/java/com/bobocode/blyznytsia/bibernate/exception/BibernateException.java (hunk 5 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/exception/BibernateException.java:8]: '+

File: src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/AnnotatedSampleEntity.java (hunk 8 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/AnnotatedSampleEntity.java]: `+`

File: src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/EntityWithNoId.java (hunk 9 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/EntityWithNoId.java]: makes sense, +

File: src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/EntityWithNoId.java (hunk 10 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/EntityWithNoId.java]: `+`

File: src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/SampleEntity.java (hunk 11 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/testdata/entity/SampleEntity.java]: `+`

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java (hunk 12 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java]: `+`

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java (hunk 13 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java:16]: `+`

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/EntityUtilTest.java (hunk 15 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/util/EntityUtilTest.java]: `+`

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/EntityUtilTest.java (hunk 16 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/util/EntityUtilTest.java]: `+`

File: lombok.config (hunk 17 above)
Comment (olehhaliak) [reply lombok.config]: `+`

File: src/main/java/com/bobocode/blyznytsia/bibernate/persister/EntityPersister.java (hunk 6 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/persister/EntityPersister.java]: Agreed, I can't find a scenario where the return type will actually be used. Updated to void.

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java (hunk 14 above)
Comment (olehhaliak) [reply src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java:8]: Good idea. Currently, I'm heavily refactoring test cases, and I will include this change with the next PR

File: src/main/java/com/bobocode/blyznytsia/bibernate/exception/BibernateException.java (hunk 4 above)
Comment (mvoronov) [reply src/main/java/com/bobocode/blyznytsia/bibernate/exception/BibernateException.java:3]: Ok, fine with me

File: src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Column.java (hunk 1 above)
Comment (mvoronov) [reply src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Column.java:8]: Ok, I see

File: src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Table.java (hunk 3 above)
Comment (mvoronov) [reply src/main/java/com/bobocode/blyznytsia/bibernate/annotation/Table.java:11]: It's a bit inconvenient, but we probably could improve it later on

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java (hunk 14 above)
Comment (mvoronov) [reply src/test/java/com/bobocode/blyznytsia/bibernate/util/CaseUtilTest.java:8]: Ok

File: src/main/java/com/bobocode/blyznytsia/bibernate/util/EntityUtil.java (hunk 7 above)
Comment (olehhaliak) [reply src/main/java/com/bobocode/blyznytsia/bibernate/util/EntityUtil.java]: From my perspective, managing composite keys at this stage will be excessive. Added an extra check for such a scenario.

File: src/test/java/com/bobocode/blyznytsia/bibernate/util/EntityUtilTest.java [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+  @Test
+  void resolveEntityIdFieldShouldThrowMalformedEntityExceptionWhenNoIdFound() {
+    assertThrows(MalformedEntityException.class, () -> EntityUtil.resolveEntityIdField(EntityWithoutId.class));
+  }
+
+  @Test
+  void resolveEntityIdFieldShouldThrowMalformedEntityExceptionWhenMultipleIdFound() {
+    assertThrows(MalformedEntityException.class, () -> EntityUtil.resolveEntityIdField(EntityWithMultipleIds.class));
+  }
```
Comment (mvoronov) [inline src/test/java/com/bobocode/blyznytsia/bibernate/util/EntityUtilTest.java:52]: In these two cases MalformedEntityException is thrown, so it would be more precise also to check an error message

**Review by mvoronov — CHANGES_REQUESTED** [review]

**Review by marynamelnichuk — APPROVED** [review]

**Review by mvoronov — APPROVED** [review]
